### PR TITLE
ARROW-1773: [C++] Add casts from date/time types to compatible signed integers

### PR DESCRIPTION
**Pull request by Licht-T**
This closes [ARROW-1773](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-1773).

### CODE REVIEW

File: python/pyarrow/tests/test_array.py
Comment (wesm) [inline python/pyarrow/tests/test_array.py]:
```
In [3]: pa.array(np.array([1,2,3], dtype='i4'), type='time32[s]')
Out[3]: 
<pyarrow.lib.Time32Array object at 0x7ffb0a5164a8>
[
  datetime.time(0, 0, 1),
  datetime.time(0, 0, 2),
  datetime.time(0, 0, 3)
]
```

File: python/pyarrow/tests/test_array.py
Comment (wesm) [inline python/pyarrow/tests/test_array.py]: same as above

```
In [4]: pa.array(np.array([1,2,3], dtype='i8'), type='date64')
Out[4]: 
<pyarrow.lib.Date64Array object at 0x7ffb0a516868>
[
  datetime.date(1970, 1, 1),
  datetime.date(1970, 1, 1),
  datetime.date(1970, 1, 1)
]
```

File: python/pyarrow/tests/test_array.py
Comment (wesm) [inline python/pyarrow/tests/test_array.py]: We might want a couple of tests in C++ to verify that this is indeed zero copy

**Comment by wesm** [comment]
Once this patch is ready, please do not merge. My work on ARROW-1559 has become a bit entangled with this, so I will take responsibility for rebasing this patch after

**Review by xhochy — CHANGES_REQUESTED** [review]
Marking this as DO NOT MERGE, just to have a visual clue to not merge it ;)

**Review by wesm — APPROVED** [review]
+1. added a couple test cases per comments

**Comment by wesm** [comment]
Casts from timestamp to int64 aren't implemented, so let me add these

**Comment by wesm** [comment]
OK, this can be merged now on green build

**Comment by Licht-T** [comment]
Thanks @wesm!

**Pull request closed without merging**
